### PR TITLE
explain: deprecate `EXPLAIN VIEW`

### PR DESCRIPTION
**Pull request by aalexandrov**
This PR deprecates the `EXPLAIN VIEW` code path and refactors code.

### Motivation

* This PR adds a known-desirable feature.

Part 3 of 4 from the tasks listed in #20652. 
Design doc PR: #20941.

### Tips for reviewer

- Deprecate `EXPLAIN VIEW` statement.
- Split code paths for `EXPLAIN [...] PLAN` and `EXPLAIN TIMESTAMP`.
- Remove obsolete `mz_sql_parser::ast::statement::ExplainOptions`.
- Move `mz_repr::explain::Explainee` to `plan::Explainee`.
- Move `raw_plan` and `row_set_finishing` parameters from `ExplainPlan` to `Explainee::Query`.
- Remove `Explainee::Dataflow` variant.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by philip-stoev — APPROVED** [review]
testdrive changes OK

File: src/adapter/src/coord/sequencer/inner.rs [hunk 1]
```diff
@@ -2858,32 +2846,30 @@ impl Coordinator {
...
             explainee,
         } = plan;
 
-        assert_ne!(stage, ExplainStage::Timestamp);
-        assert!(matches!(
-            explainee,
-            Explainee::Dataflow(_) | Explainee::Query
-        ));
+        let Explainee::Query { raw_plan, row_set_finishing } = explainee else {
+            // This is currently asserted in the `sequence_explain` code that
```
Comment (ggevay) [inline src/adapter/src/coord/sequencer/inner.rs:2869]: `sequence_explain_plan`

File: src/adapter/src/coord/sequencer/inner.rs [hunk 2]
```diff
@@ -2729,53 +2728,42 @@ impl Coordinator {
...
             format,
             config,
             no_errors: _,
             explainee,
         } = plan;
 
-        assert!(row_set_finishing.is_none());
-
         let Explainee::MaterializedView(id) = explainee else {
             // This is currently asserted in the `sequence_explain` code that
```
Comment (ggevay) [inline src/adapter/src/coord/sequencer/inner.rs:2768]: `sequence_explain_plan`

File: src/sql/src/plan/statement/dml.rs [hunk 3]
```diff
@@ -250,123 +244,131 @@ pub fn describe_explain(
...
+            config.filter_pushdown = scx.catalog.system_vars().persist_stats_filter_enabled();
+        }
+
+        config
+    };
+
+    let explainee = match explainee {
+        Explainee::View(_) => {
+            bail_never_supported!(
+                "EXPLAIN ... VIEW <view_name>",
```
Comment (ggevay) [inline src/sql/src/plan/statement/dml.rs:295]: The problem with this suggestion is that if the view is indexed, then I'll just get a `ReadExistingIndex`, and no real plan. I'm not sure what would be a better suggestion. Maybe suggest calling EXPLAIN on an index if the view is indexed. Or maybe suggest calling EXPLAIN with the view's definition.

File: src/sql-parser/src/ast/defs/statement.rs [hunk 4]
```diff
@@ -2933,8 +2952,7 @@ impl<T: AstInfo> AstDisplay for Assignment<T> {
 }
 impl_display_t!(Assignment);
 
-/// Specifies what [Statement::Explain] is actually explaining
-/// The new API
+/// Specifies what [Statement::ExplainPlan] is actually explaining The new API
```
Comment (ggevay) [inline src/sql-parser/src/ast/defs/statement.rs:2955]: "The new API" is probably some stale comment fragment.

File: src/sql-parser/src/parser.rs [hunk 5]
```diff
@@ -6689,17 +6687,32 @@ impl<'a> Parser<'a> {
...
+            self.parse_explain_timestamp()
+                .map_parser_err(StatementKind::ExplainTimestamp)
+        } else {
+            self.parse_explain_plan()
+                .map_parser_err(StatementKind::ExplainPlan)
+        }
+    }
+
+    /// Parse an `EXPLAIN` statement, assuming that the `EXPLAIN` token
+    /// has already been consumed.
```
Comment (ggevay) [inline src/sql-parser/src/parser.rs:6701]: copy-pasted from `parse_explain`

File: src/sql/src/plan/statement/dml.rs [hunk 6]
```diff
@@ -250,123 +244,131 @@ pub fn describe_explain(
...
         mz_sql_parser::ast::ExplainFormat::Dot => ExplainFormat::Dot,
     };
 
-    Ok(Plan::Explain(ExplainPlan {
-        raw_plan: expr,
-        row_set_finishing: finishing,
-        stage,
+    let raw_plan = {
+        // Previously we would bail here for ORDER BY and LIMIT; this has been relaxed to silently
+        // report the plan without the ORDER BY and LIMIT decorations (which are done in post).
```
Comment (ggevay) [inline src/sql/src/plan/statement/dml.rs:357]: Is this comment still needed?

File: test/sqllogictest/transform/topk.slt [hunk 7]
```diff
@@ -105,10 +106,12 @@ create default index on v1
...
 
 query T multiline
-explain with(arity, join_impls) view v1;
+explain with(arity, join_impls) select * from v1;
 ----
-materialize.public.v1:
-  TopK order_by=[#0 asc nulls_last] limit=3 offset=3 // { arity: 2 }
-    Get materialize.public.t1 // { arity: 2 }
+Explained Query (fast path):
+  ReadExistingIndex materialize.public.v1_primary_idx
```
Comment (ggevay) [inline test/sqllogictest/transform/topk.slt:112]: This demonstrates the problem with the suggestion of `select * from viewname`. We are unfortunately no longer testing the real plan here.

File: test/sqllogictest/transform/topk.slt [hunk 8]
```diff
@@ -56,13 +56,14 @@ statement ok
...
+EXPLAIN WITH(arity, join_impls) SELECT * FROM plan_test1
 ----
-materialize.public.plan_test1:
-  Project (#4, #0, #1) // { arity: 3 }
-    Map ((bigint_to_numeric(#2) / bigint_to_numeric(case when (#3 = 0) then null else #3 end))) // { arity: 5 }
-      Reduce group_by=[((#0 + #1) + #2), ((#0 + #1) + #3)] aggregates=[sum(#3), count(#3)] // { arity: 4 }
-        Get materialize.public.test1 // { arity: 4 }
+Explained Query (fast path):
+  Project (#2, #0, #1)
+    ReadExistingIndex materialize.public.plan_test1_primary_idx
```
Comment (ggevay) [inline test/sqllogictest/transform/topk.slt:63]: same here

File: test/sqllogictest/explain/physical_plan_as_text.slt [hunk 9]
```diff
@@ -167,10 +183,10 @@ EOF
 # TopKBasic plan.
 query T multiline
 EXPLAIN PHYSICAL PLAN WITH(no_fast_path) AS TEXT FOR
-VIEW ov
+SELECT * FROM ov
 ----
-materialize.public.ov:
-  TopK::Basic order_by=[#1 asc nulls_last, #0 desc nulls_first] limit=5
+Explained Query:
+  TopK::MonotonicTopK order_by=[#1 asc nulls_last, #0 desc nulls_first] limit=5 must_consolidate
```
Comment (ggevay) [inline test/sqllogictest/explain/physical_plan_as_text.slt:189]: This no longer matches the comment at the top of the test.

**Review by ggevay — COMMENTED** [review]
Still going through it, but some minor comments:

**Review by ggevay — APPROVED** [review]
Looks good, apart from the minor comments above.

File: src/sql/src/plan/statement/dml.rs (hunk 3 above)
Comment (aalexandrov) [reply src/sql/src/plan/statement/dml.rs:295]: > Maybe suggest calling EXPLAIN on an index if the view is indexed.

OK, I will add this suggestion when I add support for `EXPLAIN INDEX`.

File: src/sql/src/plan/statement/dml.rs (hunk 6 above)
Comment (aalexandrov) [reply src/sql/src/plan/statement/dml.rs:357]: Removed this.

File: test/sqllogictest/transform/topk.slt (hunk 8 above)
Comment (aalexandrov) [reply test/sqllogictest/transform/topk.slt:63]: See comment above.

File: src/adapter/src/coord/sequencer/inner.rs [hunk 10]
```diff
@@ -3001,17 +2985,9 @@ impl Coordinator {
...
-                (id, false, cluster_id)
-            }
-            Explainee::Query => {
-                let cluster_id = catalog.resolve_target_cluster(target_cluster, session)?.id;
-                (GlobalId::Explain, true, cluster_id)
-            }
-            _ => unreachable!(),
-        };
+        let cluster_id = catalog.resolve_target_cluster(target_cluster, session)?.id;
+        let explainee_id = GlobalId::Explain;
```
Comment (ParkMyCar) [inline src/adapter/src/coord/sequencer/inner.rs:2989]: nit: can you leave a comment on why we choose these defaults?

**Review by ParkMyCar — APPROVED** [review]
This looks good to me! I have two thoughts though:

1. Let's make sure we update Devex so they know of the syntax change
2. Is it possible someone has a View or Mat View using the now deprecated `EXPLAIN` syntax? I really doubt anyone does, but it's something that came to mind

**Comment by aalexandrov** [comment]
> 1. Let's make sure we update Devex so they know of the syntax change

Will do.

> 2. Is it possible someone has a View or Mat View using the now deprecated `EXPLAIN` syntax? I really doubt anyone does, but it's something that came to mind

No, `EXPLAIN` statements cannot be part of catalog item definitions.

**Comment by ParkMyCar** [comment]
> Will do.

Thank you!

> No, `EXPLAIN` statements cannot be part of catalog item definitions.

Perfect 🎉

**Comment by aalexandrov** [comment]
@morsapaes: I'll tag you next week when I'll have a separate PR that splits the documentation pages for `EXPLAIN PLAN` and `EXPLAIN TIMESTAMP`.

**Comment by aalexandrov** [comment]
I don't wanna block this on the Nightlies build - this PR does not introduce risky changes (apart for deprecating syntax which is not marked as backwards compatible), so any Nightlies failures due to that can be hopefully fixed asynchronously next week.

**Pull request merged**